### ISSUE TITLE
Type annotation guidelines

### ISSUE THREAD

**Issue by Illviljan** [pydata/xarray#7050]
### What is your issue?

Dask has a pretty nice guidelines for type hinting, see https://github.com/dask/community/issues/255.

Notable for us is to avoid adding typing in docstrings to avoid duplicating information.

**Comment by headtr1ck** [comment]
How exactly does it work with the automatic types in the docstring?

**Comment by benbovy** [comment]
> How exactly does it work with the automatic types in the docstring?

Yes, what I really like in the html API documentation is the types (and especially the hyperlinks!) next to each name in the parameter list. Does napoleon support generating those links automatically from the typed signature?

<img width="681" alt="Screenshot 2022-09-21 at 13 00 04" src="https://user-images.githubusercontent.com/4160723/191487662-2e5b6b02-9b59-45f2-80f2-610207ed779e.png">
